daemon/graphdriver/overlay2/overlay.go

```diff
@@ -536,8 +540,27 @@ func (d *Driver) Exists(id string) bool {
 	return err == nil
 }
 
+// isParent returns if the passed in parent is the direct parent of the passed in layer
+func (d *Driver) isParent(id, parent string) bool {
+	lowers, err := d.getLowerDirs(id)
+	if err != nil {
+		return false
+	}
+	parentDir := d.dir(parent)
+	var ld string
+	if len(lowers) > 0 {
+		ld = filepath.Dir(lowers[0])
+	}
+	return ld == parentDir
+}
+
 // ApplyDiff applies the new layer into a root
 func (d *Driver) ApplyDiff(id string, parent string, diff io.Reader) (size int64, err error) {
+	if parent != "" && !d.isParent(id, parent) {
+		logrus.Debugf("using naive diff for ApplyDiff id: %s, parent: %s", id, parent)
```

Comment:
I'm not a fan of all the debug messages in the graph drivers.  Our debug logs are already super verbose without adding much in a production setting.

This is useful for development on this feature but I think we should remove them now